internal/protocol/connector.go

```diff
@@ -126,56 +128,100 @@ func (c *Connector) connectAttemptAll(ctx context.Context, log logging.Func) (*P
-			return protocol, nil
-		}
-		if leader == "" {
-			// This server does not know who the current leader is,
-			// try with the next target.
-			log(logging.Warn, "no known leader")
-			continue
-		}
+			if ctx.Err() != nil {
+				return
+			}
 
-		// If we get here, it means this server reported that another
-		// server is the leader, let's close the connection to this
-		// server and try with the suggested one.
-		log(logging.Debug, "connect to reported leader %s", leader)
+			log := func(l logging.Level, format string, a ...interface{}) {
+				format = fmt.Sprintf("server %s: ", server.Address) + format
+				log(l, format, a...)
+			}
 
-		ctx, cancel = context.WithTimeout(ctx, c.config.AttemptTimeout)
-		defer cancel()
+			ctx, cancel := context.WithTimeout(ctx, c.config.AttemptTimeout)
+			defer cancel()
 
-		protocol, _, err = c.connectAttemptOne(ctx, leader, log)
-		if err != nil {
-			// The leader reported by the previous server is
-			// unavailable, try with the next target.
-			log(logging.Warn, "reported leader unavailable err=%v", err)
-			continue
-		}
-		if protocol == nil {
-			// The leader reported by the target server does not consider itself
-			// the leader, try with the next target.
-			log(logging.Warn, "reported leader server is not the leader")
+			protocol, leader, err := c.connectAttemptOne(ctx, server.Address, log)
+			if err != nil {
+				// This server is unavailable, try with the next target.
+				log(logging.Warn, err.Error())
+				return
+			}
+			if protocol != nil {
+				// We found the leader
+				log(logging.Debug, "connected")
+				pc <- protocol
+				return
+			}
+			if leader == "" {
+				// This server does not know who the current leader is,
+				// try with the next target.
+				log(logging.Warn, "no known leader")
+				return
+			}
+
+			// If we get here, it means this server reported that another
+			// server is the leader, let's close the connection to this
+			// server and try with the suggested one.
+			log(logging.Debug, "connect to reported leader %s", leader)
+
+			ctx, cancel = context.WithTimeout(ctx, c.config.AttemptTimeout)
+			defer cancel()
+
+			protocol, _, err = c.connectAttemptOne(ctx, leader, log)
+			if err != nil {
+				// The leader reported by the previous server is
+				// unavailable, try with the next target.
+				log(logging.Warn, "reported leader unavailable err=%v", err)
+				return
+			}
+			if protocol == nil {
+				// The leader reported by the target server does not consider itself
+				// the leader, try with the next target.
+				log(logging.Warn, "reported leader server is not the leader")
+				return
+			}
+			log(logging.Debug, "connected")
+			pc <- protocol
+		}(server, protocolChan)
+	}
+
+	// Read from protocol chan, cancel context\
+	var protocol *Protocol
+	for pc := range protocolChan {
+		if protocol != nil {
+			pc.Close()
 			continue
 		}
-		log(logging.Debug, "connected")
+		protocol = pc
+		cancel()
+	}
+
+	if protocol != nil {
```

Comment:
I feel this would be a bit clearer as follows:

```go
protocol, ok := <-protocolChan
if !ok {
    return nil, ErrNoAvailableLeader
}
cancel()
for extra := range protocolChan {
    extra.Close()
}
return protocol, nil
```

Reply:
I like it